### PR TITLE
Fragments are leaked

### PR DESCRIPTION
**Pull request by jrw972**
Problem
-------

Fragments are stored indefinitely.

Solution
--------

Expire fragments 300 seconds after the first fragment.

Note
----

A better design would:

1. Check that a fragment is from an associated writer before it is
stored.

2. Purge the fragments from a writer when it goes away.

### CODE REVIEW

File: dds/DCPS/transport/framework/TransportReassembly.h [hunk 1]
```diff
@@ -97,17 +108,22 @@ class OpenDDS_Dcps_Export TransportReassembly {
   struct FragInfo {
     FragInfo()
       : complete_(false), have_first_(false), range_list_(), total_frags_(0) {}
```
Comment (simpsont-oci) [inline dds/DCPS/transport/framework/TransportReassembly.h:110]: Do you need to initialize expiration_ here?

File: dds/DCPS/transport/framework/TransportReassembly.cpp [hunk 2]
```diff
@@ -270,17 +270,28 @@ TransportReassembly::reassemble_i(const SequenceRange& seqRange,
       data.header_.sequence_.getValue(), OPENDDS_STRING(conv).c_str()));
   }
 
+  // Purge out expired fragments.
+  const MonotonicTimePoint now = MonotonicTimePoint::now();
+  for (ExpirationQueue::iterator pos = expiration_queue_.begin(), limit = expiration_queue_.end();
```
Comment (simpsont-oci) [inline dds/DCPS/transport/framework/TransportReassembly.cpp:275]: use upper_limit(now)?

File: dds/DCPS/transport/framework/TransportReassembly.cpp [hunk 3]
```diff
@@ -373,7 +384,23 @@ TransportReassembly::data_unavailable(const SequenceNumber& dataSampleSeq,
...
-  fragments_.erase(key);
+  erase_i(fragments_.find(key));
+}
+
+void
+TransportReassembly::erase_i(FragInfoMap::iterator pos)
+{
+  if (pos != fragments_.end()) {
+    const FragKey& key = pos->first;
+    ExpirationQueue::iterator pos2 = expiration_queue_.find(pos->second.expiration_);
```
Comment (simpsont-oci) [inline dds/DCPS/transport/framework/TransportReassembly.cpp:395]: I think lower_limit() and upper_limit() would help here?

File: dds/DCPS/transport/framework/TransportReassembly.cpp [hunk 4]
```diff
@@ -270,17 +270,28 @@ TransportReassembly::reassemble_i(const SequenceRange& seqRange,
...
+    const MonotonicTimePoint expiration = MonotonicTimePoint::now() + TimeDuration(300);
+    fragments_[key] = FragInfo(firstFrag, FragRangeList(1, FragRange(seqRange, data)), total_frags, expiration);
+    expiration_queue_.insert(std::make_pair(expiration, key));
     // since this is the first fragment we've seen, it can't possibly be done
     VDBG((LM_DEBUG, "(%P|%t) DBG:   TransportReassembly::reassemble() "
       "stored first frag, returning false (incomplete)\n"));
     return false;
   } else {
     if (iter->second.complete_) {
+      // TODO: Why is this returning false?
```
Comment (simpsont-oci) [inline dds/DCPS/transport/framework/TransportReassembly.cpp:294]: We only return true when there's something to process. If it's been completed already, someone else has already gotten the "true".

File: dds/DCPS/transport/framework/TransportReassembly.cpp (hunk 3 above)
Comment (mitza-oci) [reply dds/DCPS/transport/framework/TransportReassembly.cpp:395]: With multimap, find() is not guaranteed to give us the first match.  I'll change on my branch to using equal_range.

File: dds/DCPS/transport/framework/TransportReassembly.cpp (hunk 4 above)
Comment (mitza-oci) [reply dds/DCPS/transport/framework/TransportReassembly.cpp:294]: I updated the comment on my branch

File: dds/DCPS/transport/framework/TransportReassembly.cpp (hunk 2 above)
Comment (mitza-oci) [reply dds/DCPS/transport/framework/TransportReassembly.cpp:275]: Updated on my branch

File: dds/DCPS/transport/framework/TransportReassembly.h (hunk 1 above)
Comment (mitza-oci) [reply dds/DCPS/transport/framework/TransportReassembly.h:110]: I think it's OK defaulted to 0.  Any time we have a default constructed FragInfo it's immediately overwritten by an assignment from a FragInfo will all details.

**Pull request closed without merging**